torch/nn/attention/_fa4.py

```diff
@@ -0,0 +1,599 @@
+    window_size_right: int | None,
+    seqused_k: torch.Tensor | None,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    module = _FA4_STATE.module
+    if module is None:
+        raise RuntimeError("FA4 module not loaded")
+    dense = cum_seq_q is None
+    if dense:
+        q = _fa4_prepare_dense(query)
+        k = _fa4_prepare_dense(key)
+        v = _fa4_prepare_dense(value)
+    else:
+        q = query.contiguous()
+        k = key.contiguous()
+        v = value.contiguous()
+    kwargs: dict[str, Any] = {
+        "softmax_scale": scale,
+        "causal": is_causal,
+        "window_size_left": _fa4_to_optional_int(window_size_left),
+        "window_size_right": _fa4_to_optional_int(window_size_right),
+        "return_lse": True,
+        "cu_seqlens_q": _fa4_as_int32(cum_seq_q),
+        "cu_seqlens_k": _fa4_as_int32(cum_seq_k),
+        "seqused_k": seqused_k.contiguous() if seqused_k is not None else None,
+    }
+    out, lse = module._flash_attn_fwd(q, k, v, **kwargs)
+    if dense:
+        out = _fa4_restore_dense(out)
+    return out, lse.contiguous()
+
+
+def _fa4_run_backward(
+    grad_out: torch.Tensor,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    out: torch.Tensor,
+    logsumexp: torch.Tensor,
+    cum_seq_q: torch.Tensor | None,
+    cum_seq_k: torch.Tensor | None,
+    scale: float | None,
+    is_causal: bool,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    module = _FA4_STATE.module
+    if module is None:
+        raise RuntimeError("FA4 module not loaded")
+    dense = cum_seq_q is None
+    if dense:
+        q = _fa4_prepare_dense(query)
+        k = _fa4_prepare_dense(key)
+        v = _fa4_prepare_dense(value)
+        o = _fa4_prepare_dense(out)
+        do = _fa4_prepare_dense(grad_out)
+    else:
+        q = query.contiguous()
+        k = key.contiguous()
+        v = value.contiguous()
+        o = out.contiguous()
+        do = grad_out.contiguous()
+    dq, dk, dv = module._flash_attn_bwd(
+        q,
+        k,
+        v,
+        o,
+        do,
+        logsumexp.contiguous(),
+        softmax_scale=scale,
+        causal=is_causal,
+        cu_seqlens_q=_fa4_as_int32(cum_seq_q),
+        cu_seqlens_k=_fa4_as_int32(cum_seq_k),
+    )
+    if dense:
+        dq = _fa4_restore_dense(dq)
+        dk = _fa4_restore_dense(dk)
+        dv = _fa4_restore_dense(dv)
+    return dq, dk, dv
+
+
+def _fa4_flash_attention_forward_impl(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    cum_seq_q: torch.Tensor | None,
+    cum_seq_k: torch.Tensor | None,
+    max_q: int,
+    max_k: int,
+    dropout_p: float,
+    is_causal: bool,
+    return_debug_mask: bool,
+    *,
+    scale: float | None = None,
+    window_size_left: int | None = None,
+    window_size_right: int | None = None,
+    seqused_k: torch.Tensor | None = None,
+    alibi_slopes: torch.Tensor | None = None,
+):
+    if not _FA4_STATE.registered:
```

Comment:
I'm confused how that can happen haha
The only way to get here is if this op is registered to the dispatcher